File: src/eyedb/RecMode.cc
/* 
   EyeDB Object Database Management System
   Copyright (C) 1994-2018 SYSRA
   
   EyeDB is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.
   
   EyeDB is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.
   
   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA 
*/

/*
   Author: Eric Viara <viara@sysra.com>
*/


#include "eyedb_p.h"

namespace eyedb {

  const RecMode *FullRecurs;
  const RecMode *NoRecurs;

  const RecMode *RecMode::FullRecurs;
  const RecMode *RecMode::NoRecurs;

  void RecMode::init(void)
  {
    static const RecMode FullRecurs_rec(RecMode_FullRecurs);;
    static const RecMode NoRecurs_rec(RecMode_NoRecurs);;

    eyedb::FullRecurs = &FullRecurs_rec;
    eyedb::NoRecurs = &NoRecurs_rec;

    FullRecurs = eyedb::FullRecurs;
    NoRecurs = eyedb::NoRecurs;
  }

  void RecMode::_release(void)
  {
  }

  RecMode::RecMode(RecModeType typ)
  {
    type = typ;
    u.pred = 0;
    u.fnm.fnm_cnt = 0;
  }

  RecMode::RecMode(Bool (*p)(const Object *,
			     const Attribute *, int))
  {
    type = RecMode_Predicat;
    u.pred = p;
  }

#define SET_FNAMES(FIRST, AP) \
  va_list AP; \
  va_start(AP, FIRST); \
  char *s; \
  u.fnm.fnm_cnt = 1; \
 \
  while (s = va_arg(AP, char *)) \
    { \
      u.fnm.fnm_cnt++; \
    } \
  u.fnm.fnm = (char **)malloc(u.fnm.fnm_cnt * sizeof(char *)); \
  va_start(AP, FIRST); \
  int k = 0; \
  u.fnm.fnm[k++] = strdup(first); \
  while (s = va_arg(AP, char *)) \
    { \
      u.fnm.fnm[k++] = strdup(s); \
    } \
  va_end(AP);

  RecMode::RecMode(const char *first, ...)
  {
    type = RecMode_FieldNames;
    u.fnm._not = False;

    SET_FNAMES(first, ap);
  }

  RecMode::RecMode(Bool _not, const char *first, ...)
  {
    type = RecMode_FieldNames;
    u.fnm._not = _not;

    SET_FNAMES(first, ap);
  }

  Bool RecMode::isAgregRecurs(const Attribute *agreg, int n, const Object *o) const
  {
    if (type == RecMode_NoRecurs)
      return False;

    if (type == RecMode_FullRecurs)
      return True;

    if (type == RecMode_Predicat)
      return u.pred(o, agreg, n);

    if (type == RecMode_FieldNames)
      {
	const char *name = agreg->getName();
	for (int i = 0; i < u.fnm.fnm_cnt; i++)
	  if (!strcmp(name, u.fnm.fnm[i]))
	    return True;
     
	return False;
      }

    return False;
  }

}

File: src/java/src/org/eyedb/RecMode.java
/* 
   EyeDB Object Database Management System
   Copyright (C) 1994-2008 SYSRA
   
   EyeDB is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.
   
   EyeDB is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.
   
   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA 
*/

/*
   Author: Eric Viara <viara@sysra.com>
*/

package org.eyedb;

public class RecMode {

  static final int RecMode_NoRecurs   = 0;
  static final int RecMode_FullRecurs = 1;
  static final int RecMode_FieldNames = 2;
  static final int RecMode_Predicat   = 3;

  static public RecMode FullRecurs = new RecMode(RecMode_FullRecurs);
  static public RecMode NoRecurs   = new RecMode(RecMode_NoRecurs);

  public RecMode(int type) {
    this.type = type;
  }

  public RecMode(String s) {
    this.type = RecMode_FieldNames;
  }

  public RecMode(boolean b, String s) {
    this.type = RecMode_FieldNames;
  }

  public int getType() {return type;}

  public boolean isAgregRecurs(Attribute agr) throws Exception {
    return isAgregRecurs(agr, -1, null);
  }

  public boolean isAgregRecurs(Attribute agr, int j) throws Exception {
    return isAgregRecurs(agr, j, null);
  }

  public boolean isAgregRecurs(Attribute agr, int j, Object o)
    throws Exception {

    if (type == RecMode_FullRecurs)
      return true;

    if (type == RecMode_NoRecurs)
      return false;

    if (type == RecMode_Predicat)
	throw new Exception("INTERNAL_ERROR",
			       "RecMode_Predicat not implemented");

    if (type == RecMode_FieldNames)
	throw new Exception("INTERNAL_ERROR",
			       "RecMode_FieldNames not implemented");

    return false;
  }

  private int type;
}
